src/distributed-transaction/distributed-algorithms.md

```diff
@@ -0,0 +1,54 @@
+# Distributed Algorithms
+
+## Two-Phase Commit
+
+In transaction processing, databases, and computer networking, the two-phase commit protocol (2PC) is a type of atomic commitment protocol (ACP). It is a distributed algorithm that coordinates all the processes that participate in a distributed atomic transaction on whether to commit or abort (rollback) the transaction (it is a specialized type of consensus protocol). The protocol achieves its goal even in many cases of temporary system failure (involving either process, network node, communication, etc. failures), and is thus widely used. However, it is not resilient to all possible failure configurations, and in rare cases, user (e.g., a system's administrator) intervention is needed to remedy an outcome. To accommodate recovery from failure (automatic in most cases) the protocol's participants use logging of the protocol's states. Log records, which are typically slow to generate but survive failures, are used by the protocol's recovery procedures. Many protocol variants exist that primarily differ in logging strategies and recovery mechanisms. Though usually intended to be used infrequently, recovery procedures compose a substantial portion of the protocol, due to many possible failure scenarios to be considered and supported by the protocol.
+
+### Basic Algorithm of 2PC
+
+#### prepare phase
+
+The coordinator sends a prepare message to all cohorts and waits until it has received a reply from all cohorts.
+
+#### commit phase
+
+If the coordinator received an agreement message from all cohorts during the commit-request phase, the coordinator sends a commit message to all the cohorts.
+
+If any cohort votes No during the prepare phase (or the coordinator's timeout expires), the coordinator sends a rollback message to all the cohorts.
+
+### Disadvantages of 2PC
+
+The greatest disadvantage of the two-phase commit protocol is that it is a blocking protocol. If the coordinator fails permanently, some cohorts will never resolve their transactions: After a cohort has sent an agreement message to the coordinator, it will block until a commit or rollback is received.
+
+### 2PC Practice in TiKV
+
+In TiKV we adopt percolator transaction model which is a variant of two phase commit(we will describe this transaction model deeply later). To address the disadvantage of coordinator fails permanently, percolator doesn’t use any node as coordinator, instead it uses one of the transaction involved keys as coordinator, which we call it primary key, and the other keys we call them secondary keys. Since each key has multiple replicas, and data is keep consistent between these replicas by using consensus protocol(raft in TiKV), one node’s failure doesn’t affect the accessibility of data. So percolator can tolerate node fails permanently.
```

Comment:
`In TiKV we adopted the [Percolator transaction model](https://storage.googleapis.com/pub-tools-public-publication-data/pdf/36726.pdf)` so our readers can learn more if they want.